File: activity/20200805/junhaolin/15.3Sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        set<vector<int>> result;
        map<int,int> count;
        for (int i=0; i<nums.size(); i++)
            count[nums[i]]++;
        
        vector<int> newNums;
        for (const auto& [key, value] : count) 
            newNums.insert(newNums.end(), value > 3 ? 3 : value, key);
        
        for (int i=0; i<newNums.size(); i++) {
            for (int j=i+1; j<newNums.size(); j++) {
                int c = 0 - (newNums[i]+newNums[j]);
                if (count.find(c) != count.end()) {
                    if ((c == newNums[i] && c == newNums[j] && count[c]>=3)
                       || (c == newNums[i] && c != newNums[j] && count[c]>=2)
                       || (c != newNums[i] && c == newNums[j] && count[c]>=2) 
                       || (c != newNums[i] && c != newNums[j])) {
                        vector<int> triplet = vector<int> {newNums[i], newNums[j], c};
                        sort(triplet.begin(), triplet.end());
                        result.insert(triplet);
                    }
                }
            }
        }
        return vector<vector<int>> (result.begin(), result.end());
    }
};

File: activity/20200805/ninezero/15_3sum.go
func threeSum(nums []int) [][]int {
	// a + b = -c 排序加双指针
	var res [][]int
	if len(nums) < 3 {
		return res
	}
	sort.Ints(nums)
	// 从最小开始循环到==0结束
	for i := 0; i < len(nums) && nums[i] <= 0; i++ {
		// 避免重复
		if i >= 1 && nums[i] == nums[i-1] {
			continue
		}
		left, right := i+1, len(nums)-1
		for left < right {
			if nums[left] + nums[right] + nums[i] == 0 {
				res = append(res, []int{nums[i], nums[left], nums[right]})
				left += 1
				right -= 1
				// 去重
				for left < right && nums[left] == nums[left-1] {
					left++
				}
				for left < right && nums[right] == nums[right+1] {
					right--
				}
			} else if nums[left] + nums[right] + nums[i] < 0 {
				left += 1
			} else {
				right -= 1
			}
		}
	}
	return res
}
